core/test/utils/unsort_matrix.hpp

```diff
@@ -0,0 +1,140 @@
+namespace test {
+
+
+// Plan for now: shuffle values and column indices to unsort the given matrix
+// without changing the represented matrix.
+template <typename ValueType, typename IndexType, typename RandomEngine>
+void unsort_matrix(matrix::Csr<ValueType, IndexType> *mtx,
+                   RandomEngine &&engine)
+{
+    using value_type = ValueType;
+    using index_type = IndexType;
+    auto size = mtx->get_size();
+    if (mtx->get_num_stored_elements() <= 0) {
+        return;
+    }
+    const auto &exec = mtx->get_executor();
+    const auto &master = exec->get_master();
+
+    // If exec is not the master/host, extract the master and perform the
+    // unsorting there, followed by copying it back
+    if (exec != master) {
+        auto h_mtx = mtx->clone(master);
+        unsort_matrix(lend(h_mtx), engine);
+        mtx->copy_from(lend(h_mtx));
+        return;
+    }
+
+    auto vals = mtx->get_values();
+    auto row_ptrs = mtx->get_row_ptrs();
+    auto cols = mtx->get_col_idxs();
+
+    for (index_type row = 0; row < size[0]; ++row) {
+        auto start = row_ptrs[row];
+        auto end = row_ptrs[row + 1];
+        auto iterator = gko::detail::IteratorFactory<IndexType, ValueType>(
+            cols + start, vals + start, end - start);
+        std::shuffle(iterator.begin(), iterator.end(), engine);
+    }
+}
+
+
+// Plan for now: shuffle values and column indices to unsort the given matrix
+// without changing the represented matrix.
+// Note: This expects COO to be properly sorted by row index (which is required
+//       by the Ginkgo COO format)
+template <typename ValueType, typename IndexType, typename RandomEngine>
+void unsort_matrix(matrix::Coo<ValueType, IndexType> *mtx,
+                   RandomEngine &&engine)
+{
+    using value_type = ValueType;
+    using index_type = IndexType;
+    auto nnz = mtx->get_num_stored_elements();
+    if (nnz <= 0) {
+        return;
+    }
+
+    const auto &exec = mtx->get_executor();
+    const auto &master = exec->get_master();
+
+    // If exec is not the master/host, extract the master and perform the
+    // unsorting there, followed by copying it back
+    if (exec != master) {
+        auto h_mtx = mtx->clone(master);
+        unsort_matrix(lend(h_mtx), engine);
+        mtx->copy_from(lend(h_mtx));
+        return;
+    }
+
+    auto vals = mtx->get_values();
+    auto rows = mtx->get_row_idxs();
+    auto cols = mtx->get_col_idxs();
+
+    auto current_row = rows[0];
+    for (IndexType i = 0; i < nnz;) {
+        auto start = i;
+        while (i < nnz && rows[i] == current_row) {
+            ++i;
+        }
+        current_row = rows[i];
+        auto end = i;
+        auto iterator = gko::detail::IteratorFactory<IndexType, ValueType>(
+            cols + start, vals + start, end - start);
+        // since the row entries are supposed to be the same, there is no need
+        // to swap
+        std::shuffle(iterator.begin(), iterator.end(), engine);
+    }
```

Comment:
We could maybe simplify this into a std::shuffle followed by a std::stable_sort by only the first index? Might make the intent and correctness clearer?

Reply:
Good idea, then we don't rely on the partially sorting of COO.

Reply:
I changed it a bit since I can't sort 3 values in different places according to just one of them (the self-written iterator helps only for 2 values at the same time).
Since performance is not important here (it is only for tests), I simply convert it to `matrix_data`, do the shuffling and sorting, and then write it to the matrix again.